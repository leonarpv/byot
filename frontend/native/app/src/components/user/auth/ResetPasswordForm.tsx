import * as React from 'react';
import {Text, View} from 'react-native-ui-lib';
import {Formik} from 'formik';
import {useTranslation} from 'react-i18next';
import {IResetPassword} from '@byot-frontend/common/src/types/interfaces/IResetPassword';
import {ResetPassword} from '@byot-frontend/common/src/types/dto/ResetPassword';
import {resetPasswordSchema} from '@byot-frontend/common/src/types/schemas/validation/ResetPasswordSchema';
import {TextField} from '../../elements/lib/TextField';
import {Button} from '../../elements/lib/Button';

export const testIDs = {
  newPassword: 'resetPassword-newPassword',
  passwordRepeat: 'resetPassword-passwordRepeat',
  submit: 'resetPassword-submit',
};

interface Props {
  loading: boolean;
  onSubmit: (values: IResetPassword) => void;
  token: string;
}

export const ResetPasswordForm: React.FC<Props> = (props: Props) => {
  const {t} = useTranslation();
  const [password, setPassword] = React.useState('');
  const initialValues: IResetPassword = new ResetPassword({token: props.token});
  return (
    <>
      <Text heading>{t('Reset password')}</Text>
      <Formik
        initialValues={initialValues}
        validateOnBlur
        validateOnChange
        validationSchema={resetPasswordSchema(t, password)}
        onSubmit={values => props.onSubmit(values)}>
        {({values, handleChange, handleSubmit, handleBlur, errors, touched}) => (
          <>
            <View marginB-10>
              <TextField
                placeholder={t('Enter new password')}
                value={values.newPassword}
                secureTextEntry
                onBlur={handleBlur('newPassword')}
                onChangeText={text => {
                  setPassword(text);
                  handleChange('newPassword')(text);
                }}
                infoText
                error={(touched.newPassword && errors.newPassword) as string}
                testID={testIDs.newPassword}
                accessibilityLabel={testIDs.newPassword}
                autoCapitalize="none"
              />
            </View>
            <View marginB-10>
              <TextField
                placeholder={t('Confirm password')}
                value={values.passwordRepeat}
                secureTextEntry
                onBlur={handleBlur('passwordRepeat')}
                onChangeText={handleChange('passwordRepeat')}
                infoText
                error={(touched.passwordRepeat && errors.passwordRepeat) as string}
                testID={testIDs.passwordRepeat}
                accessibilityLabel={testIDs.passwordRepeat}
                autoCapitalize="none"
              />
            </View>
            <View marginB-10>
              <Button
                label={t('Reset password')}
                testID={testIDs.submit}
                onPress={handleSubmit}
                loading={props.loading}
              />
            </View>
          </>
        )}
      </Formik>
    </>
  );
};
