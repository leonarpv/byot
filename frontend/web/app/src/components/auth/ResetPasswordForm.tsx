import * as React from 'react';
import {Grid} from '@material-ui/core';
import {TFunction} from 'i18next';
import {Form, Formik} from 'formik';
import * as Yup from 'yup';
import {PlainLayoutNarrow} from '@byot-frontend/web-common/src/components/plain-layout/PlainLayoutNarrow';
import {PlainLayoutTitle} from '@byot-frontend/web-common/src/components/plain-layout/PlainLayoutTitle';
import {useTranslation} from 'react-i18next';
import {Input} from '@byot-frontend/web-common/src/components/elementary/form/Input';
import {Button} from '@byot-frontend/web-common/src/components/elementary/form/Button';
import {IResetPassword} from '@byot-frontend/common/src/types/interfaces/IResetPassword';
import {ResetPassword} from '@byot-frontend/common/src/types/dto/ResetPassword';

const resetPassSchema = (t: TFunction, password: string) =>
  Yup.object().shape({
    newPassword: Yup.string()
      .required(t('Enter password'))
      .min(8, t('Password must be longer than 8 characters'))
      .matches(/[^A-Za-z0-9]+/, t('Password must contain at least one special character')),
    passwordRepeat: Yup.string()
      .required(t('Confirm password'))
      .oneOf([password], t('Passwords are not the same')),
  });

interface Props {
  loading: boolean;
  onSubmit: (values: IResetPassword) => void;
  token: string;
}

export const ResetPasswordForm: React.FC<Props> = (props: Props) => {
  const {t} = useTranslation();
  const [password, setPassword] = React.useState('');
  const initialValues: IResetPassword = new ResetPassword({token: props.token});
  return (
    <PlainLayoutNarrow>
      <PlainLayoutTitle>{t('Reset password')}</PlainLayoutTitle>
      <Formik
        initialValues={initialValues}
        validateOnBlur
        validateOnChange
        validationSchema={resetPassSchema(t, password)}
        onSubmit={values => props.onSubmit(values)}>
        {({values, handleChange, handleBlur, errors, touched}) => (
          <Form data-testid="common-auth-resetPassword-form-form">
            <Grid container spacing={2}>
              <Grid item xs={12}>
                <Input
                  data-testid="common-auth-resetPassword-form-newPassword"
                  label={t('Enter new password')}
                  error={!!(touched.newPassword && errors.newPassword)}
                  helperText={touched.newPassword && errors.newPassword}
                  onChange={e => {
                    setPassword(e.target.value);
                    handleChange('newPassword')(e);
                  }}
                  onBlur={handleBlur('newPassword')}
                  value={values.newPassword}
                  type="password"
                />
              </Grid>
              <Grid item xs={12}>
                <Input
                  data-testid="common-auth-resetPassword-form-passwordRepeat"
                  label={t('Confirm password')}
                  error={!!(touched.passwordRepeat && errors.passwordRepeat)}
                  helperText={touched.passwordRepeat && errors.passwordRepeat}
                  onChange={handleChange('passwordRepeat')}
                  onBlur={handleBlur('passwordRepeat')}
                  value={values.passwordRepeat}
                  type="password"
                />
              </Grid>
              <Grid item xs={12}>
                <Button type="submit" fullWidth loading={props.loading}>
                  {t('Reset password')}
                </Button>
              </Grid>
            </Grid>
          </Form>
        )}
      </Formik>
    </PlainLayoutNarrow>
  );
};
