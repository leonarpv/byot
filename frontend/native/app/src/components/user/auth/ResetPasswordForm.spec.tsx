import * as React from 'react';
import {render, fireEvent, wait} from '@testing-library/react-native';
import {ResetPasswordForm, testIDs} from './ResetPasswordForm';

describe('<ResetPasswordForm/>', () => {
  it('should successfully submit change', () => {
    const onSubmit = jest.fn();
    const {asJSON, getByTestId} = render(
      <ResetPasswordForm onSubmit={onSubmit} loading={false} token="token" />
    );
    const pass = 'pass-123A';
    fireEvent.changeText(getByTestId(testIDs.newPassword), pass);
    fireEvent.changeText(getByTestId(testIDs.passwordRepeat), pass);
    wait(() =>
      expect(onSubmit).toBeCalledWith({
        token: 'token',
        newPassword: pass,
        passwordRepeat: pass,
      })
    );
    expect(asJSON()).toMatchSnapshot();
  });
  it('should fail submitting because of non-matching passwords', () => {
    const onSubmit = jest.fn();
    const {asJSON, getByTestId} = render(
      <ResetPasswordForm onSubmit={onSubmit} loading={false} token="token" />
    );
    const pass = 'pass-123A';
    fireEvent.changeText(getByTestId(testIDs.newPassword), pass);
    fireEvent.changeText(getByTestId(testIDs.passwordRepeat), pass + 'addition');
    wait(() => expect(onSubmit).not.toBeCalled());
    expect(asJSON()).toMatchSnapshot();
  });
  it('should fail submitting because of easy passwords', () => {
    const onSubmit = jest.fn();
    const {asJSON, getByTestId} = render(
      <ResetPasswordForm onSubmit={onSubmit} loading={false} token="token" />
    );
    const pass = 'pass';
    fireEvent.changeText(getByTestId(testIDs.newPassword), pass);
    fireEvent.changeText(getByTestId(testIDs.passwordRepeat), pass);
    wait(() => expect(onSubmit).not.toBeCalled());
    expect(asJSON()).toMatchSnapshot();
  });
});
