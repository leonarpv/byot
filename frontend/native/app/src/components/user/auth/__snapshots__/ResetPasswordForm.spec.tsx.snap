// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResetPasswordForm/> should fail submitting because of easy passwords 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Text
    heading={true}
  >
    Reset password
  </Text>
  <View
    marginB-10={true}
  >
    <View
      paddingB-8={true}
      paddingL-12={true}
      paddingR-12={true}
      paddingT-8={true}
      style={
        Array [
          Object {
            "backgroundColor": "#FAFAF1",
            "borderRadius": 8,
          },
          undefined,
        ]
      }
    >
      <TouchableWithoutFeedback>
        <TextInput
          accessibilityLabel="resetPassword-newPassword"
          allowFontScaling={true}
          autoCapitalize="none"
          color="#FAFAF1"
          enableErrors={false}
          floatOnFocus={true}
          floatingPlaceholder={true}
          floatingPlaceholderColor={
            Object {
              "default": "#FAFAF1",
              "error": "#FAFAF1",
              "focus": "#FAFAF1",
            }
          }
          floatingPlaceholderStyle={Object {}}
          hideUnderline={true}
          infoText={true}
          placeholder="Enter new password"
          rejectResponderTermination={true}
          secureTextEntry={true}
          style={
            Array [
              undefined,
              Object {
                "marginBottom": 4,
                "marginTop": -4,
              },
            ]
          }
          testID="resetPassword-newPassword"
          underlineColorAndroid="transparent"
          value="pass"
        />
      </TouchableWithoutFeedback>
    </View>
    <Text
      accessibilityLabel="textField-helperText"
      error={false}
      style={
        Object {
          "fontSize": 12,
          "marginTop": 2,
          "paddingLeft": 12,
        }
      }
      success={false}
      testID="textField-helperText"
    >
       
    </Text>
  </View>
  <View
    marginB-10={true}
  >
    <View
      paddingB-8={true}
      paddingL-12={true}
      paddingR-12={true}
      paddingT-8={true}
      style={
        Array [
          Object {
            "backgroundColor": "#FAFAF1",
            "borderRadius": 8,
          },
          undefined,
        ]
      }
    >
      <TouchableWithoutFeedback>
        <TextInput
          accessibilityLabel="resetPassword-passwordRepeat"
          allowFontScaling={true}
          autoCapitalize="none"
          color="#FAFAF1"
          enableErrors={false}
          floatOnFocus={true}
          floatingPlaceholder={true}
          floatingPlaceholderColor={
            Object {
              "default": "#FAFAF1",
              "error": "#FAFAF1",
              "focus": "#FAFAF1",
            }
          }
          floatingPlaceholderStyle={Object {}}
          hideUnderline={true}
          infoText={true}
          placeholder="Confirm password"
          rejectResponderTermination={true}
          secureTextEntry={true}
          style={
            Array [
              undefined,
              Object {
                "marginBottom": 4,
                "marginTop": -4,
              },
            ]
          }
          testID="resetPassword-passwordRepeat"
          underlineColorAndroid="transparent"
          value="pass"
        />
      </TouchableWithoutFeedback>
    </View>
    <Text
      accessibilityLabel="textField-helperText"
      error={false}
      style={
        Object {
          "fontSize": 12,
          "marginTop": 2,
          "paddingLeft": 12,
        }
      }
      success={false}
      testID="textField-helperText"
    >
       
    </Text>
  </View>
  <View
    marginB-10={true}
  >
    <TouchableOpacity
      accessibilityLabel="resetPassword-submit"
      label="Reset password"
      loading={false}
      testID="resetPassword-submit"
    />
  </View>
</View>
`;

exports[`<ResetPasswordForm/> should fail submitting because of non-matching passwords 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Text
    heading={true}
  >
    Reset password
  </Text>
  <View
    marginB-10={true}
  >
    <View
      paddingB-8={true}
      paddingL-12={true}
      paddingR-12={true}
      paddingT-8={true}
      style={
        Array [
          Object {
            "backgroundColor": "#FAFAF1",
            "borderRadius": 8,
          },
          undefined,
        ]
      }
    >
      <TouchableWithoutFeedback>
        <TextInput
          accessibilityLabel="resetPassword-newPassword"
          allowFontScaling={true}
          autoCapitalize="none"
          color="#FAFAF1"
          enableErrors={false}
          floatOnFocus={true}
          floatingPlaceholder={true}
          floatingPlaceholderColor={
            Object {
              "default": "#FAFAF1",
              "error": "#FAFAF1",
              "focus": "#FAFAF1",
            }
          }
          floatingPlaceholderStyle={Object {}}
          hideUnderline={true}
          infoText={true}
          placeholder="Enter new password"
          rejectResponderTermination={true}
          secureTextEntry={true}
          style={
            Array [
              undefined,
              Object {
                "marginBottom": 4,
                "marginTop": -4,
              },
            ]
          }
          testID="resetPassword-newPassword"
          underlineColorAndroid="transparent"
          value="pass-123A"
        />
      </TouchableWithoutFeedback>
    </View>
    <Text
      accessibilityLabel="textField-helperText"
      error={false}
      style={
        Object {
          "fontSize": 12,
          "marginTop": 2,
          "paddingLeft": 12,
        }
      }
      success={false}
      testID="textField-helperText"
    >
       
    </Text>
  </View>
  <View
    marginB-10={true}
  >
    <View
      paddingB-8={true}
      paddingL-12={true}
      paddingR-12={true}
      paddingT-8={true}
      style={
        Array [
          Object {
            "backgroundColor": "#FAFAF1",
            "borderRadius": 8,
          },
          undefined,
        ]
      }
    >
      <TouchableWithoutFeedback>
        <TextInput
          accessibilityLabel="resetPassword-passwordRepeat"
          allowFontScaling={true}
          autoCapitalize="none"
          color="#FAFAF1"
          enableErrors={false}
          floatOnFocus={true}
          floatingPlaceholder={true}
          floatingPlaceholderColor={
            Object {
              "default": "#FAFAF1",
              "error": "#FAFAF1",
              "focus": "#FAFAF1",
            }
          }
          floatingPlaceholderStyle={Object {}}
          hideUnderline={true}
          infoText={true}
          placeholder="Confirm password"
          rejectResponderTermination={true}
          secureTextEntry={true}
          style={
            Array [
              undefined,
              Object {
                "marginBottom": 4,
                "marginTop": -4,
              },
            ]
          }
          testID="resetPassword-passwordRepeat"
          underlineColorAndroid="transparent"
          value="pass-123Aaddition"
        />
      </TouchableWithoutFeedback>
    </View>
    <Text
      accessibilityLabel="textField-helperText"
      error={false}
      style={
        Object {
          "fontSize": 12,
          "marginTop": 2,
          "paddingLeft": 12,
        }
      }
      success={false}
      testID="textField-helperText"
    >
       
    </Text>
  </View>
  <View
    marginB-10={true}
  >
    <TouchableOpacity
      accessibilityLabel="resetPassword-submit"
      label="Reset password"
      loading={false}
      testID="resetPassword-submit"
    />
  </View>
</View>
`;

exports[`<ResetPasswordForm/> should successfully submit change 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Text
    heading={true}
  >
    Reset password
  </Text>
  <View
    marginB-10={true}
  >
    <View
      paddingB-8={true}
      paddingL-12={true}
      paddingR-12={true}
      paddingT-8={true}
      style={
        Array [
          Object {
            "backgroundColor": "#FAFAF1",
            "borderRadius": 8,
          },
          undefined,
        ]
      }
    >
      <TouchableWithoutFeedback>
        <TextInput
          accessibilityLabel="resetPassword-newPassword"
          allowFontScaling={true}
          autoCapitalize="none"
          color="#FAFAF1"
          enableErrors={false}
          floatOnFocus={true}
          floatingPlaceholder={true}
          floatingPlaceholderColor={
            Object {
              "default": "#FAFAF1",
              "error": "#FAFAF1",
              "focus": "#FAFAF1",
            }
          }
          floatingPlaceholderStyle={Object {}}
          hideUnderline={true}
          infoText={true}
          placeholder="Enter new password"
          rejectResponderTermination={true}
          secureTextEntry={true}
          style={
            Array [
              undefined,
              Object {
                "marginBottom": 4,
                "marginTop": -4,
              },
            ]
          }
          testID="resetPassword-newPassword"
          underlineColorAndroid="transparent"
          value="pass-123A"
        />
      </TouchableWithoutFeedback>
    </View>
    <Text
      accessibilityLabel="textField-helperText"
      error={false}
      style={
        Object {
          "fontSize": 12,
          "marginTop": 2,
          "paddingLeft": 12,
        }
      }
      success={false}
      testID="textField-helperText"
    >
       
    </Text>
  </View>
  <View
    marginB-10={true}
  >
    <View
      paddingB-8={true}
      paddingL-12={true}
      paddingR-12={true}
      paddingT-8={true}
      style={
        Array [
          Object {
            "backgroundColor": "#FAFAF1",
            "borderRadius": 8,
          },
          undefined,
        ]
      }
    >
      <TouchableWithoutFeedback>
        <TextInput
          accessibilityLabel="resetPassword-passwordRepeat"
          allowFontScaling={true}
          autoCapitalize="none"
          color="#FAFAF1"
          enableErrors={false}
          floatOnFocus={true}
          floatingPlaceholder={true}
          floatingPlaceholderColor={
            Object {
              "default": "#FAFAF1",
              "error": "#FAFAF1",
              "focus": "#FAFAF1",
            }
          }
          floatingPlaceholderStyle={Object {}}
          hideUnderline={true}
          infoText={true}
          placeholder="Confirm password"
          rejectResponderTermination={true}
          secureTextEntry={true}
          style={
            Array [
              undefined,
              Object {
                "marginBottom": 4,
                "marginTop": -4,
              },
            ]
          }
          testID="resetPassword-passwordRepeat"
          underlineColorAndroid="transparent"
          value="pass-123A"
        />
      </TouchableWithoutFeedback>
    </View>
    <Text
      accessibilityLabel="textField-helperText"
      error={false}
      style={
        Object {
          "fontSize": 12,
          "marginTop": 2,
          "paddingLeft": 12,
        }
      }
      success={false}
      testID="textField-helperText"
    >
       
    </Text>
  </View>
  <View
    marginB-10={true}
  >
    <TouchableOpacity
      accessibilityLabel="resetPassword-submit"
      label="Reset password"
      loading={false}
      testID="resetPassword-submit"
    />
  </View>
</View>
`;
